include/soci/type-holder.h

```diff
@@ -44,49 +50,310 @@ T* checked_ptr_cast(U* ptr)
     return static_cast<T*>(ptr);
 }
 
-// Base class holder + derived class type_holder for storing type data
-// instances in a container of holder objects
-template <typename T>
-class type_holder;
-
-class holder
+// Type safe conversion that fails at compilation if instantiated
+template <typename T, typename U, typename Enable = void>
+struct soci_cast
 {
-public:
-    holder() {}
-    virtual ~holder() {}
+    static inline T cast(U)
+    {
+        throw std::bad_cast();
+    }
+};
 
-    template<typename T>
-    T get()
+// Type safe conversion that is a noop
+template <typename T, typename U>
+struct soci_cast<
+    T, U,
+    typename std::enable_if<std::is_same<T, U>::value>::type>
+{
+    static inline T cast(U val)
     {
-        type_holder<T>* p = checked_ptr_cast<type_holder<T> >(this);
-        if (p)
-        {
-            return p->template value<T>();
-        }
-        else
+        return val;
+    }
+};
+
+// Type safe conversion that is widening the type
+template <typename T, typename U>
+struct soci_cast<
+    T, U,
+    typename std::enable_if<(
+        !std::is_same<T, U>::value &&
+        std::is_integral<T>::value &&
+        std::is_integral<U>::value
+    )>::type>
+{
+    static inline T cast(U val) {
+        intmax_t t_min = static_cast<intmax_t>((std::numeric_limits<T>::min)());
+        intmax_t u_min = static_cast<intmax_t>((std::numeric_limits<U>::min)());
+        uintmax_t t_max = static_cast<uintmax_t>((std::numeric_limits<T>::max)());
+        uintmax_t u_max = static_cast<uintmax_t>((std::numeric_limits<U>::max)());
+
+        if ((t_min > u_min && val < static_cast<U>(t_min)) ||
+            (t_max < u_max && val > static_cast<U>(t_max)))
         {
             throw std::bad_cast();
         }
-    }
 
-private:
+        return static_cast<T>(val);
+    }
+};
 
-    template<typename T>
-    T value();
+union type_holder
+{
+    std::string* s;
+    int8_t* i8;
+    int16_t* i16;
+    int32_t* i32;
+    int64_t* i64;
+    uint8_t* u8;
+    uint16_t* u16;
+    uint32_t* u32;
+    uint64_t* u64;
+    double* d;
+    std::tm* t;
 };
 
 template <typename T>
-class type_holder : public holder
+struct type_holder_trait
+{
+    static_assert(std::is_same<T, void>::value, "Unmatched raw type");
+    // dummy value to satisfy the template engine, never used
+    static const db_type type = (db_type)0;
+};
```

Comment:
Do we actually have to define it? I.e. could we just declare it without defining it?